### ISSUE TITLE
[FR]: Context Menu: "Speak from here"

### ISSUE THREAD

**Issue by Luram123** [ImperialSquid/zotero-zotts#119]
### Checklist

- [X] I have used the search function for [**OPEN**](https://github.com/ImperialSquid/zotero-zotts/issues) issues to see if someone else has already submitted the same feature request.
- [X] I have **also** used the search function for [**CLOSED**](https://github.com/ImperialSquid/zotero-zotts/issues?q=is%3Aissue+is%3Aclosed) issues to see if the feature was already implemented and is just waiting to be released, or if the feature was rejected.
- [X] I have read through the [changelog](https://github.com/ImperialSquid/zotero-zotts/blob/main/docs/CHANGELOG.md) to confirm the feature doesn't already exist.
- [X] I have read through the [roadmap](https://github.com/ImperialSquid/zotero-zotts/blob/main/docs/ROADMAP.md) to confirm the feature isn't already planned. (Note - you can still open a feature request if you think the feature should have a higher priority/should be implemented differently/etc!)
- [X] I have described the feature request with as much detail as possible.
- [X] This request contains only one single feature, **not** a list of multiple (related) features.

### Feature or improvement you want

I've found something similar in the roadmap which is called _[implement skipping](https://github.com/ImperialSquid/zotero-zotts/blob/main/docs/ROADMAP.md#implement-skipping)_. But what I have in mind is not the same.
In my opinion it is an essential feature of a TTS plugin that one can just right-click at one spot within the text and "speak from here" in the context menu. So that the voice starts reading from that point on.

### Why should this be added?

Up until now the plugin can read aloud PDF, but either from the beginning of the whole thing, or just one selected passage. That way, when you want to start a PDF in the middle, you have to select each page again and again once the TTS is done reading the last selection.

**Comment by ImperialSquid** [comment]
Sounds like a good feature idea, I'll add it to my todo list! 👌 

Just as an fyi though, it might not happen super soon. 

---

(Technical details) There are various limitations when it comes to accessing the text of a paper due to security built into Firefox/Zotero (the full text and selected text are both easy to get since they're stored in specific places outside of the paper's internal structure, but accessing the raw text elements isn't), so unless the user were to select enough text to uniquely identify their position in the full text, it would be difficult to know what text element was being selected and thus where to start reading from...

Implementing highlighting will also require accessing the raw text elements though and that's a massively requested feature, so I'll look into adding this one at the same time as I do that.

**Comment by diabolusss** [comment]
> (Technical details) There are various limitations when it comes to accessing the text of a paper due to security built into Firefox/Zotero (the full text and selected text are both easy to get since they're stored in specific places outside of the paper's internal structure, but accessing the raw text elements isn't), so unless the user were to select enough text to uniquely identify their position in the full text, it would be difficult to know what text element was being selected and thus where to start reading from...

I'm wondering if it might be easier to implement this feature in two stages. For the first stage, we could use a workaround that involves selecting enough text to uniquely identify the user's position in the full text, and then offer an option like "Continue reading from here (experimental)" or simply "Read from here (experimental)". The final stage would then be the fully implemented version. What do you think?

**Comment by ImperialSquid** [comment]
Yeah that was something I considered but was resistant towards actually doing at the time. When this was first raised the plugin was only a few months into being released and I didn't want to release a half arsed version of the feature if I could get it perfect with a bit more effort (and since it's tied to highlighting text which is like **_the_** feature a lot of people want added, I thought I could knock out two birds with one stone).

But given it's been a while (and I've had little response from the Zotero devs about highlighting stuff which was my priority at the time), and I agree this would be a very useful feature _especially_ for long papers or platforms like Linux where pausing is iffy for some users, I think this will be a feature I focus on in the near future 👌

Given its somewhat experimental nature I'll probably release it as a beta feature first before doing a proper release to everyone (only for a few weeks and just to iron out any major oversights).

**Comment by diabolusss** [comment]
> Yeah that was something I considered but was resistant towards actually doing at the time. When this was first raised the plugin was only a few months into being released and I didn't want to release a half arsed version of the feature if I could get it perfect with a bit more effort (and since it's tied to highlighting text which is like **_the_** feature a lot of people want added, I thought I could knock out two birds with one stone).
> 
> But given it's been a while (and I've had little response from the Zotero devs about highlighting stuff which was my priority at the time), and I agree this would be a very useful feature _especially_ for long papers or platforms like Linux where pausing is iffy for some users, I think this will be a feature I focus on in the near future 👌
> 
> Given its somewhat experimental nature I'll probably release it as a beta feature first before doing a proper release to everyone (only for a few weeks and just to iron out any major oversights).

Thanks for sharing your thoughts and plans on this feature! I completely understand the urge to perfect it before a full release—I've been there too. Life often has its own way of shifting priorities, and I really appreciate the dedication you're putting in, especially when things get busy. Zotero's unique capabilities make it such an essential tool for me — not only for academic papers but also for book reading—and I'm genuinely grateful for all your hard work (and yes, Piper’s sweet voices definitely add a special touch!).

I’d love to contribute in some way, though my free time is a way limited these days. If you’re open to monetary donations to help support further development, please let us know how we can chip in.

Thanks again for everything, and I look forward to seeing the beta release! Of course, at your own pace. ;)

**Comment by ImperialSquid** [comment]
> I’d love to contribute in some way, though my free time is a way limited these days. If you’re open to monetary donations to help support further development, please let us know how we can chip in.

Thanks, that would be really appreciated!

I've actually been quietly thinking about adding links for that for a while (I also wasn't sure how worth doing it was since ZoTTS is a very small and very new plugin compared to others so I didn't know if there would be interest).

> I'm genuinely grateful for all your hard work

> Thanks again for everything

Cheers, I appreciate the encouragement!

**Comment by diabolusss** [comment]
> (I also wasn't sure how worth doing it was since ZoTTS is a very small and very new plugin compared to others so I didn't know if there would be interest).

Just fyi, I'm a big fan of the "I'm Feeling Lucky" button (that I believe Google invented, but that's not important). ;D

**Comment by jrob95** [comment]
Just to pile on, and I'm not sure if this is a separate issue, but would it be possible to add highlighting the word being spoken (and I hate to even say this, because TTS is word is awful) similar to the way TTS works in MS word?

Thanks, Jack
p.s. big fan of this package, keep up the awesome work!

**Comment by ImperialSquid** [comment]
Hi @jrob95, thanks for the comment

Strictly speaking it's a separate thing but I'll answer about it here anyway lol.

And here's everything as it currently stands in full, just for future readers:

Trust me, I'm very aware of the desire for highlighting text lol... It's one of the things I _most_ wanted to do before the initial release which I didn't manage to, and since then it comes up very often in feature requests (which is partly why I wrote up the ROADMAP document, to share my thoughts about common feature requests).

Unfortunately, implementing highlighting is much more complex than anything else ZoTTS does at the moment, since currently everything ZoTTS needs to do can be handled with existing functionality in Zotero.

But highlighting text on my end isn't something I can do currently, so adding highlighting to ZoTTS requires _first adding highlighting to **Zotero**_ (ie highlighting that isn't caused by the user selecting text, but can be done by the program itself). This is unfortunately fairly complex since Zotero is a heavily modified version of Firefox, so you can imagine there's a lot in the way of privacy and protection around documents you might be reading, making adding highlighting on the program difficult.

As I'm sure you can imagine, the Zotero devs are pretty busy on the day to day putting in features and putting out fires lol, so something like this has understandably kinda fallen by the way side with them (especially since ZoTTS released at the same time as Zotero 7).

And unfortunately, as much as I want ZoTTS to succeed and put loads of effort into it, my day job is done with Python and R, languages like JavaScript/TypeScript, which Zotero and ZoTTS use, are very much **not** my area of expertise... (Honestly, ZoTTS started out as a hobby project, I never thought it would get more than a dozen users, let alone the 7.5k or so it currently has!) So, implementing highlighting in Zotero myself would be an impossibly huge undertaking...

Hopefully that explains a little about why a feature like this is still yet to be added? (I hope you enjoyed your peek behind the curtain of software dev lol)

Trust me, no one more than me wants highlighting added to ZoTTS, in my mind it's the one feature it needs to be "complete" (not that I would stop after that, just that it's a critically missing component). But there are several major blockages in the way of getting it added, and no easy fixes to any of them unfortunately...

**Comment by jrob95** [comment]
Yeah, thanks. I totally understand, might be something we can raise with the ?core zotero team in their forum.

Thanks for the detailed repsonse, and good luck and godspeed with the package, I have used for a few days, and even lacking all of these "nice to haves" it is very functional and easy to use.

All the best mate, 
Jack

**Comment by ImperialSquid** [comment]
@Luram123 @rusak47 @jrob95 @dmeiser

Speak From Here has now been implemented 🎉

From the release notes:
- Added the ability to "speak from here" in a paper
  - ZoTTS can only "speak from here" with some selected text to start from, you'll see a notification if the selected start point couldn't be used for some reason (no text selected, text is not unique, position couldn't be identified, etc)
  - The "Speak From Here" menu item when you make a selection will start speaking from the selected text to the end
  - By default, `Ctrl/Cmd + S` will speak the selected text and `Ctrl/Cmd + Shift + S` will "speak from here"
    - You can swap these under the Shift Modifier portion of the Shortcuts section in the preferences menu
  - The UI play button in the top right of the reader will always just speak the selected text (though I'm happy to change this if people want)

Using the selected text isn't perfect (sometimes there are mismatches with the full text Zotero has stored) so if you run into a lot of issues, or it doesn't seem to work on some papers, please open an issue and let me know!

ZoTTS should update itself if you have Zotero open so you don't need to do anything, if you want to trigger a manual update anyway you can do so from `Tools > Plugins > ZoTTS > Gear Icon > Check for Updates`

**Comment by ImperialSquid** [comment]
Oh also @rusak47, I added links to support me through GitHub Sponsors, but only if you're still interested and no pressure! 😁

**Comment by rusak47** [comment]
That's fantastic news! 🎉 Thank you, Ben, for bringing it to a life!

I’ve tested it on a few papers and love how seamlessly it starts (_most of the time_)! 

There is a tiny hiccup with page transitions for short documents - it speaks to the end of the page and then speech-dispatcher freezes, along with ZoTTs. The only way to resolve this is by restarting both the speech-dispatcher service and Zotero.

So far i've tested these documents:
 - Artificial Intelligence: A Modern Approach (3rd Edition) by Stuart Russell and Peter Norvig, more than 1000 pages - TTS is completely silent and freezes instantly. _(Obviously, it's not reasonable and wasn't intended for this, but I was eager to try it anyway.)_
 - [Can Language Models Help in System Security? Investigating Log Anomaly Detection using BERT by Crispin Almodovar et al., 9 pages](https://www.researchgate.net/publication/368332816_Can_Language_Models_Help_in_System_Security_Investigating_Log_Anomaly_Detection_using_BERT), [A model for bridging the gap between neuroscience and education by Jodi Tommerdahl, 14 pages](https://www.tandfonline.com/doi/abs/10.1080/03054980903518936) - manages to speak to the end of the page.

Grateful for your hard work! 🙌

P.S. Please let me know if you need any log data.

**Comment by ImperialSquid** [comment]
> That's fantastic news! 🎉 Thank you, Ben, for bringing it to a life!
> 
> I’ve tested it on a few papers and love how seamlessly it starts (_most of the time_)!
>
> Grateful for your hard work! 🙌

Happy to help 😁 thanks for your patience while I had stuff going on.

(And thanks for your **incredibly** generous donation, I was *very* surprised when I saw that! 😳😅💜)

---

>  - Artificial Intelligence: A Modern Approach (3rd Edition) by Stuart Russell and Peter Norvig, more than 1000 pages - TTS is completely silent and freezes instantly. _(Obviously, it's not reasonable and wasn't intended for this, but I was eager to try it anyway.)_

This is almost certainly caused by the text being very long (you'd get the same lag spike and freeze by speaking the full text, as well as just "speak from here"-ing near the start. 

The issue is being tracked in #173 and is currently being worked on 👌

(I'm just figuring out a graceful way to split the text that won't cause any issues disrupting words, shouldn't be hard, just needs some fiddling).

> There is a tiny hiccup with page transitions for short documents - it speaks to the end of the page and then speech-dispatcher freezes, along with ZoTTs. The only way to resolve this is by restarting both the speech-dispatcher service and Zotero
> 
>  - [Can Language Models Help in System Security? Investigating Log Anomaly Detection using BERT by Crispin Almodovar et al., 9 pages](https://www.researchgate.net/publication/368332816_Can_Language_Models_Help_in_System_Security_Investigating_Log_Anomaly_Detection_using_BERT), [A model for bridging the gap between neuroscience and education by Jodi Tommerdahl, 14 pages](https://www.tandfonline.com/doi/abs/10.1080/03054980903518936) - manages to speak to the end of the page.

Interesting that it speaks to the end of the page but just stops...

I can't think of an obvious reason why, but ZoTTS handles speaking on Linux very differently (to "allow" for pausing), so it could be something there... 🤔

> P.S. Please let me know if you need any log data.

If you wouldn't mind opening an issue for the 9 and 14 page ones that would be great, just so I don't forget to look into them at some point, thanks.

**Comment by rusak47** [comment]
> This is almost certainly caused by the text being very long (you'd get the same lag spike and freeze by speaking the full text, as well as just "speak from here"-ing near the start.
> 
> The issue is being tracked in [#173](https://github.com/ImperialSquid/zotero-zotts/issues/173) and is currently being worked on 👌

For sure, however, the steps used in issue [#173] do not break the speech-dispatcher on my system. ZoTTs simply says it can't speak the file because no voices are installed, but it speaks the selected part just fine after that.

p.s. Just happy to back the right thing. 💜✨
